rfcs/0005-multi-blockchain-support.md

```diff
@@ -0,0 +1,487 @@
+from the instance manager. So it will no longer have direct access to the subgraph store.
+
+The __instance manager__, __mapping runtime__ and related core components will be reused, though it
+will take considerable refactoring to move out the Ethereum-specific code.
+
+The __subgraph store__ will be reused since graphql entities behave the same for any chain. Even for
+the manifest and dynamic data source metadata we should be able to reuse the existing tables.
+
+#### Data structures
+
+A chain will define associated types such as `Block` and `Trigger` which will be produced by the
+adapters and over which the reusable components will need to be made generic. Since the `Chain`
+trait aggregates all chain-specific associated types, most of the codebase will be made generic over
+a `C: Chain` type parameter. The manifest data structure will also need to be made generic over the
+chain data source type.
+
+The `BlockPointer` type is shared across all chains, the existing `EthereumBlockPointer` will be
+renamed and its `hash` field will have the type changed from `H256` to `Bytes`.
+
+#### Chain Indexer
+
+Chains may include a built-in subgraph for block data, this is to be detailed in a future RFC. But
+this is an optional feature and does not replace the block ingestor.
+
+### Adding Support for a New Blockchain
+
+This section covers what it takes to add support for a new blockchain. Please
+note that this RFC specifies the details only to an extent that allow us to
+assess whether the direction is feasible. Names and data structure details may
+vary in the implementation.
+
+#### Graph Node
+
+Each blockchain integration in Graph Node lives in the `graph-node` repository
+under `network/<name>/`, like the already existing `network/ethereum/` support.
+
+Graph Node includes a number of new component traits and data types to enable
+multi-blockchain support, sketched below:
+
+```rust
+/// A compact representation of a block on a chain.
+struct BlockPtr {
+    pub number: u64,
+    pub hash: Bytes,
+}
+
+/// Represents a block in a particular chain. Each chain has its own
+/// implementation of the `Block` trait.
+trait Block: ToAsc {
+    fn number(&self) -> u64 { self.ptr().number }
+    fn hash(&self) -> { self.ptr().hash }
+    fn ptr(&self) -> BlockPointer;
+    fn parent_ptr(&self) -> Option<BlockPointer>;
+}
+
+trait DataSource<C: Blockchain> { /* ... */ }
+trait IngestorAdapter<C: Blockchain> { /* ... */ }
+trait TriggersAdapter<C: Blockchain> { /* ... */ }
+trait HostFns<C: Blockchain> { /* ... */ }
+
+trait BlockStream<C: Blockchain> {
+      fn new(&self, start_block: BlockPointer, data_sources: Vec<C::DataSource>, trigger_adapter: Arc<C::TriggerAdapter>) -> Result<Self, Error>;
+}
+
+/// Represents a blockchain supported by the node.
+trait Chain {
+    /// The DB schema for the `blocks` table and any custom tables.
+    const DB_SCHEMA: &'static str;
```

Comment:
In the implementation, I would like to keep this more flexible, and probably have a callback `initialize_chain_store_schema` or some such that will be called whenever a chain needs to set up its database schema.

Reply:
Good reminder but we don't need to work out the details right now so I'll keep this as-is for the RFC.

Reply:
What I was getting at is that a `Chain` has a lifecycle, and rather than giving big prominence to the fact that it might want to have special things happening in the store, it might be better to have method stubs in this trait that calls out the fact that there is a lifecycle, and that the rest of the system will call certain things at certain points in the lifecycle. At a minimum, there will be a call when the chain is first created, which is when the chain would create custom schema, and there might well also be a need for a call when `graph-node` starts up and wants to get the chain ready for business.

Reply:
I've added an `fn start()` method to the chain trait.